sdk/internal/testframework/request_matcher.go

```diff
@@ -0,0 +1,176 @@
+// +build go1.13
+
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package testframework
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"reflect"
+
+	"github.com/dnaeon/go-vcr/cassette"
+)
+
+type RequestMatcher struct {
+	context        TestContext
+	IgnoredHeaders map[string]*string
+	bodyMatcher    StringMatcher
+	urlMatcher     StringMatcher
+	methodMatcher  StringMatcher
+}
+
+type StringMatcher func(reqVal string, recVal string) bool
+type matcherWrapper func(matcher StringMatcher, testContext TestContext) bool
+
+var ignoredHeaders = map[string]*string{
+	"Date":                   nil,
+	"X-Ms-Date":              nil,
+	"x-ms-date":              nil,
+	"x-ms-client-request-id": nil,
+	"User-Agent":             nil,
+	"Request-Id":             nil,
+	"traceparent":            nil,
+	"Authorization":          nil,
+}
+
+const (
+	recordingHeaderMissing = "Test recording headers do not match. Header '%s' is present in request but not in recording."
+	requestHeaderMissing   = "Test recording headers do not match. Header '%s' is present in recording but not in request."
+	headerValuesMismatch   = "Test recording header '%s' does not match. request: %s, recording: %s"
+	methodMismatch         = "Test recording methods do not match. request: %s, recording: %s"
+	urlMismatch            = "Test recording URLs do not match. request: %s, recording: %s"
+	bodiesMismatch         = "Test recording bodies do not match.\nrequest: %s\nrecording: %s"
+)
+
+func DefaultMatcher(testContext TestContext) *RequestMatcher {
+	// The default sanitizer sanitizes the Authorization header
+	matcher := &RequestMatcher{
+		context:        testContext,
+		IgnoredHeaders: ignoredHeaders,
+	}
+	matcher.SetBodyMatcher(func(req string, rec string) bool {
+		return DefaultStringMatcher(req, rec)
+	})
+	matcher.SetURLMatcher(func(req string, rec string) bool {
+		return DefaultStringMatcher(req, rec)
+	})
+	matcher.SetMethodMatcher(func(req string, rec string) bool {
+		return DefaultStringMatcher(req, rec)
+	})
+
+	return matcher
+}
+
+func (m *RequestMatcher) SetBodyMatcher(matcher StringMatcher) {
+	m.bodyMatcher = func(reqVal string, recVal string) bool {
+		isMatch := matcher(reqVal, recVal)
+		if !isMatch {
+			m.context.Log(fmt.Sprintf(bodiesMismatch, recVal, recVal))
+		}
+		return isMatch
+	}
+}
+
+func (m *RequestMatcher) SetURLMatcher(matcher StringMatcher) {
+	m.urlMatcher = func(reqVal string, recVal string) bool {
+		isMatch := matcher(reqVal, recVal)
+		if !isMatch {
+			m.context.Log(fmt.Sprintf(urlMismatch, recVal, recVal))
+		}
+		return isMatch
+	}
+}
+
+func (m *RequestMatcher) SetMethodMatcher(matcher StringMatcher) {
+	m.methodMatcher = func(reqVal string, recVal string) bool {
+		isMatch := matcher(reqVal, recVal)
+		if !isMatch {
+			m.context.Log(fmt.Sprintf(methodMismatch, recVal, recVal))
+		}
+		return isMatch
+	}
+}
+
+func DefaultStringMatcher(s1 string, s2 string) bool {
```

Comment:
Can this not be exported, or do we expect users of this package to need this?

Reply:
I don't believe it needs to be exported